federation/pkg/federation-controller/ingress/ingress_controller.go

```diff
@@ -0,0 +1,360 @@
+		if err != nil {
+			glog.Errorf("Failed to get %s from %s: %v", ingress, cluster.Name, err)
+			ic.deliverIngress(ingress, 0, true)
+			return
+		}
+		desiredIngress := &extensions_v1beta1.Ingress{
+			ObjectMeta: baseIngress.ObjectMeta,
+			Spec:       baseIngress.Spec,
+		}
+
+		if !found {
+			// We can't supply server-created fields when creating a new object.
+			desiredIngress.ObjectMeta.ResourceVersion = ""
+			desiredIngress.ObjectMeta.UID = ""
+			// We always first create an ingress in the first available cluster.  Once that ingress
+			// has been created and allocated a global IP (visible via an annotation),
+			// we record that annotation on the federated ingress, and create all other cluster
+			// ingresses with that same global IP.
+			// Note: If the first cluster becomes (e.g. temporarily) unavailable, the second cluster will be allocated
+			// index 0, but eventually all ingresses will share the single global IP recorded in the annotation
+			// of the federated ingress.
+			if baseIPExists || (clusterIndex == 0) {
+				operations = append(operations, util.FederatedOperation{
+					Type:        util.OperationTypeAdd,
+					Obj:         desiredIngress,
+					ClusterName: cluster.Name,
+				})
+			}
+		} else {
+			clusterIngress := clusterIngressObj.(*extensions_v1beta1.Ingress)
+			glog.V(4).Infof("Found existing Ingress %s in cluster %s - checking if update is required", ingress, cluster.Name)
+			clusterIPName, clusterIPExists := clusterIngress.ObjectMeta.Annotations[staticIPAnnotationKey]
+			if !baseIPExists && clusterIPExists {
+				// Add annotation to federated ingress via API.
+				original, err := ic.federatedApiClient.Extensions().Ingresses(baseIngress.Namespace).Get(baseIngress.Name)
+				if err == nil {
+					original.ObjectMeta.Annotations[staticIPAnnotationKey] = clusterIPName
+					if _, err = ic.federatedApiClient.Extensions().Ingresses(baseIngress.Namespace).Update(original); err != nil {
+						glog.Errorf("Failed to add static IP annotation to federated ingress %q: %v", ingress, err)
+					}
+				} else {
+					glog.Errorf("Failed to get federated ingress %q: %v", ingress, err)
+				}
+			}
+			// Update existing ingress, if needed.
+			if !util.ObjectMetaIsEquivalent(desiredIngress.ObjectMeta, clusterIngress.ObjectMeta) ||
+				!reflect.DeepEqual(desiredIngress.Spec, clusterIngress.Spec) {
+				// TODO: In some cases Ingress controllers in the clusters add annotations, so we ideally need to exclude those from
+				// the equivalence comparison to cut down on unnecessary updates.
+				glog.V(4).Infof("Ingress %s in cluster %s needs an update: cluster ingress %v is not equivalent to federated ingress %v", ingress, cluster.Name, clusterIngress, desiredIngress)
+				// We need to use server-created fields from the cluster, not the desired object when updating.
+				desiredIngress.ObjectMeta.ResourceVersion = clusterIngress.ObjectMeta.ResourceVersion
```

Comment:
Oh no wait, I was correct.  When updating (the cluster ingress), one needs to provide the resource version and UID.